api/src/main/java/org/apache/iceberg/util/CharSequenceMap.java

```diff
@@ -132,12 +133,12 @@ public Set<CharSequence> keySet() {
       keySet.add(wrapper.get());
     }
 
-    return keySet;
+    return Collections.unmodifiableSet(keySet);
```

Comment:
A simple way to make `CharSequenceMap` obey `Map`  contract (which is super important!) is to re-use existing implementation. `TreeMap` is one that can be used for that.
The whole class can therefore be replaced with:

```java
/**
 * A map that uses char sequences as keys and compares them by value.
 *
 * @param <V> the type of values
 */
public abstract class CharSequenceMap<V> implements Map<CharSequence, V>, Serializable {
  public static <V> Map<CharSequence, V> create() {
    return new TreeMap<>(CharSequence::compare);
  }
}
```

Reply:
I submitted this as a separate PR: https://github.com/apache/iceberg/pull/11308

Reply:
Thanks @findepi. Let me check the other map implementations and make sure they follow the [Map](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Map.html) contract.

Reply:
Generally writing a good Map implementation from scratch is super hard. 
I did that for Trino once because JDK's IdentityMap violates (violated?) some Map contract aspects, and it was ton of code and even more tests. Using mainstream implementations is really much saner.

Reply:
> Generally writing a good Map implementation from scratch is super hard.

... and sometimes not worth the effort
After all, one can have map-like class without implementing Map interface  -- https://github.com/apache/iceberg/pull/11704